File: client/src/main/java/communication/ClientAuthenticator.java
package communication;

import exceptions.InvalidCredentialsException;
import exceptions.UserAlreadyExistsException;
import user.User;

import java.io.IOException;

public class ClientAuthenticator implements Authenticator {

  private final Exchanger exchanger;

  public ClientAuthenticator(Exchanger exchanger) {
    this.exchanger = exchanger;
  }

  @Override
  public void register(User user) throws UserAlreadyExistsException, IOException {
    exchanger.requestRegister(user);
    try {
      exchanger.validateRegister();
    } catch (InvalidCredentialsException e) {
      throw new UserAlreadyExistsException(user.getLogin());
    }
  }

  @Override
  public void login(User user) throws InvalidCredentialsException, IOException {
    exchanger.requestLogin(user);
    exchanger.validateLogin();
  }
}

File: client-ui/domain/src/commonMain/kotlin/ru.moleus.kollector/domain/communication/ClientAuthenticator.kt
package ru.moleus.kollector.domain.communication

import common.context.Exchanger
import common.exceptions.InvalidCredentialsException
import ru.moleus.kollector.domain.exceptions.UserAlreadyExistsException
import user.User
import java.io.IOException

class ClientAuthenticator(private val exchanger: Exchanger) : Authenticator {

    @Throws(UserAlreadyExistsException::class, IOException::class)
    override fun register(user: User) {
        try {
            exchanger.requestRegister(user)
        } catch (e: InvalidCredentialsException) {
            throw UserAlreadyExistsException(user.login)
        }
    }

    @Throws(InvalidCredentialsException::class, IOException::class)
    override fun login(user: User) {
        exchanger.requestLogin(user)
    }
}